### ISSUE TITLE
Instance for `Monoid (Event a)` ?

### ISSUE THREAD

**Issue by HeinrichApfelmus** [HeinrichApfelmus/reactive-banana#104]
The `Event` type will be changed to [no longer allow simultaneous occurrences in a single event](http://apfelmus.nfshost.com/blog/2015/08/29-frp-banana-api-redesign.html). This raises the question under which circumstances, if any, there should be an instance `Monoid (Event A)` for some types `A`.

There are various proposals and use cases:
1. A general instance `Monoid (Event a)` that forgets the right occurrence in case of simultaneity,
   
   ```
   instance Monoid (Event a) where
       mappend = unionWith (\x y -> x)
   ```
   
   I want the programmer to think about the case of simultaneous events more thoroughly, so I don't think this is a good idea.
2. The old semantics can be simulated by the type `Event [a]`. The previous `union` combinator then translates to a natural instance for list types
   
   ```
   instance Monoid (Event [a]) where
       mappend = unionWith (++)
   ```
   
   I don't know of a concrete use case, but I think this is a good idea.
3. The previous consideration suggests a very general instance
   
   ```
   instance Monoid a => Monoid (Event a) where
       mappend = unionWith mappend
   ```
   
   However, it has been [pointed out to me](https://www.reddit.com/r/haskell/comments/3ivbht/frp_api_redesign_for_reactivebanana_10_feedback/cukl16l) that this instance also works with semigroups (which lack the `mempty` element) 
   
   ```
   instance Semigroup a => Monoid (Event a) where
       mappend = unionWith (<>)
   ```
   
   This more general instance overlaps and hence is in conflict with the previous instance. More importantly, there are interesting semigroups that are not monoids, for instance [`NonEmpty a`](https://hackage.haskell.org/package/semigroups-0.16.2.2/docs/Data-List-NonEmpty.html) which describes nonempty lists. For instance, this is useful when we want to signal that a user has selected a non-zero number of elements in a list box.
4. Issue #50 discusses a possible instance
   
   ```
   instance Monoid (Event (a -> a)) where
       mappend = unionWith (flip (.))
   ```
   
   which would have nice synergies with the `accumB/E` functions. This would also overlap and hence be in conflict with general instance
   
   ```
   instance Something a => Monoid (Event a) where …
   ```
   
   It can be interpreted as the instance
   
   ```
   instance Monoid (Event (Dual (Endo a)))
   ```
   
   but with significantly less plumbing.

---

I currently towards providing exactly two instances

```
instance Monoid (Event [a])
instance Monoid (Event (a -> a))
```

and nothing else.

This way, people are free to add additional instances like

```
instance Monoid (Event (Nonempty a))
```

as needed. On the other hand, general instances like

```
instance Semigroup a => Monoid (Event a)
```

are excluded.

Thoughts? Suggestions? Do you have an example where you have used an instance like this?

**Comment by ocharles** [comment]
A problem with 

> people are free to add additional instances like

is that it requires introducing orphan instances. For an executable, fine, but if you want to use `reactive-banana` as a tool when building a library, I would be very hesistant to do that - because I make a choice that affects *all* my downstream users.

I would reject 1 because it is just too surprising. 

The dependency on `semigroups` is not that bad, but the amount of instances available is significantly less than the amount of `Monoid` instances. That said, there is a clear path to getting `semigroups` into base, and all `Monoid`s give rise to a `Semigroup`, so when that does happen you can flip a switch in assurance that you won't break anyone's code.

So my vote would be `Monoid a => Monoid (Event a)`, with a note that this will one day become `Semigroup a => Monoid (Event a)`. You may want to document/encourage the use of your two preferred instances, as with this proposal you get those instances - and more.

**Comment by HeinrichApfelmus** [comment]
> So my vote would be Monoid a => Monoid (Event a)

That was my first reaction as well, but the problem with this is that it unfortunately excludes two instances due to overlaps:

    instance Monoid (Event (NonEmpty a))
    instance Monoid (Event (a -> a))

I find the first to be very compelling (non-empty selection in a list box), for the second, I am undecided.

That's why I'm leaning towards orphan instances, because I think they're the lesser of two evils in this case.

**Comment by ocharles** [comment]
Is it not possible to flag the instance you provide as `OVERLAPPABLE` and let users chose to add these instances if they want them? As a few counters as well,

1. `instance Monoid (Event (NonEmpty a))` will be valid when you change the instance context from `Monoid a` to `Semigroup a`. In the meantime, if people don't want to add an overlapping instance, they they can use `Option (NonEmpty a)` - though of course there is the unfortunate property that that is isomorphic to `[a]` so you have to deal with an impossible case.

2. That is just `Monoid (Event (Endo a))`, and all the tools for dealing with `newtype` wrappers should be valid here. E.g., the `newtype` library, and also the Wrapped/Unwrapped combinators from `lens`.

**Comment by HeinrichApfelmus** [comment]
I'll leave this open for the future. At the moment, the function

    unions :: [Event (a -> a)] -> Event (a -> a)

which serves the most immediate `Monoid` needs.

**Comment by amigalemming** [comment]
I am in favor of
~~~~
instance Semigroup a => Monoid (Event a)
~~~~
or, second best, temporarily
~~~~
instance Monoid a => Monoid (Event a)
~~~~
as suggested by ocharles.
`Monoid (Event (NonEmptyList a))` and `Monoid (Event (a -> a))` are FlexibleInstances and risk Orphaned Instances. FlexibleInstances are a sign that there is no common pattern for the instances which usually indicates that a type class is not the right tool at all. Contrarily the Semigroup pattern matches perfectly.
Usually you regret the litte convenience caused by suchs hacks only too soon.

**Comment by avieth** [comment]
I'm glad to see this being discussed. I've been using `Event` monoidally in my work. I made a new type and accepted the minor inconvenience of constructing and deconstructing it:

```
newtype SemigroupEvent t = SemigroupEvent { runSemigroupEvent :: Event t }

instance Semigroup t => Semigroup (SemigroupEvent t) where
  SemigroupEvent left <> SemigroupEvent right = SemigroupEvent (unionWith (<>) left right)

instance Semigroup t => Monoid (SemigroupEvent t) where
  mempty = never
  mappend = (<>)
```

If a `Monoid (Event t)` is coming then I hope that `t` is not constrained to be a monoid, because then we have no monoidal expression for these:

```
unionWith const :: forall t . Event t -> Event t -> Event t
unionWith (flip const) :: forall t . Event t -> Event t -> Event t
```

But if we choose semigroup, then these correspond to the choice of [`First t`](https://hackage.haskell.org/package/semigroups-0.18.1/docs/Data-Semigroup.html#t:First) or [`Last t`](https://hackage.haskell.org/package/semigroups-0.18.1/docs/Data-Semigroup.html#t:Last) from the semigroups package.

**Comment by mitchellwrosen** [comment]
+1 to 
```haskell
instance Semigroup a => Monoid (Event a)
```

**Comment by amigalemming** [comment]
`Semigroup` is available from `base-4.9`. I think depending on the `semigroup` package is not bad, though, for users of pre-FTP versions of `base` like me. The `Semigroup` approach would automatically enable `instance Monoid (Event [a])`, `instance Monoid (Event (NonEmpty a))` and `instance Monoid (Event (Endo a))`, where the last one enables the use of `a -> a`.

**Comment by HeinrichApfelmus** [comment]
Ok, I see, `Semigroup` is in the base library since GHC 8.0. I'm sold! Thanks everyone for their continue support of the rightful cause! 😉

I will add two instances

    instance Semigroup a => Semigroup (Event a)
    instance Semigroup a => Monoid (Event a)

The [semigroups][1] package is used for compatibility with older versions of GHC. (This package does have the drawback of drawing in [TemplateHaskell][2] as a dependency, which may pose difficulties when trying to compile it with GHCJS, but that may well be a non-issue.)

  [1]: http://hackage.haskell.org/package/semigroups
  [2]: http://hackage.haskell.org/package/template-haskell

**Comment by amigalemming** [comment]
On Mon, 7 Aug 2017, Heinrich Apfelmus wrote:

> The semigroups package is used for compatibility with older versions of 
> GHC. (This package does have the drawback of drawing in TemplateHaskell 
> as a dependency, which may pose difficulties when trying to compile it 
> with GHCJS, but that may well be a non-issue.)

"semigroups" itself does not depend on template-haskell. It must be one of 
its dependencies. Can that one be disabled with one of the various package 
flags?

**Comment by HeinrichApfelmus** [comment]
> "semigroups" itself does not depend on template-haskell. It must be one of
its dependencies. Can that one be disabled with one of the various package
flags?

No idea. It appears that Template Haskell is pulled in by the `nats` package.

**Comment by mitchellwrosen** [comment]
I think this issue was closed by 1.2.0.0

**Comment by ocharles** [comment]
Yep, I think it is. Thanks for the nudge!
